internal/resources/providers/gcplib/inventory/provider.go

```diff
@@ -91,242 +75,327 @@ type dnsPolicyFields struct {
-			return displayNamesCache.Get(func() string {
-				prj, err := crmService.Projects.Get(parent).Context(ctx).Do()
-				if err != nil {
-					log.Errorf("error fetching GCP Project: %s, error: %s", parent, err)
-					return ""
-				}
-				return prj.DisplayName
-			}, parent)
-		},
-		getOrganizationDisplayName: func(ctx context.Context, parent string) string {
-			return displayNamesCache.Get(func() string {
-				org, err := crmService.Organizations.Get(parent).Context(ctx).Do()
-				if err != nil {
-					log.Errorf("error fetching GCP Org: %s, error: %s", parent, err)
-					return ""
-				}
-				return org.DisplayName
-			}, parent)
-		},
+	cloudResourceManager, err := NewResourceManagerWrapper(ctx, log, gcpConfig)
+	if err != nil {
+		return nil, err
 	}
 
 	return &Provider{
-		config:                    gcpConfig,
-		log:                       log,
-		inventory:                 assetsInventoryWrapper,
-		crm:                       crmServiceWrapper,
-		cloudAccountMetadataCache: NewMapCache[*fetching.CloudAccountMetadata](),
+		config:    gcpConfig,
+		log:       log,
+		inventory: assetsInventory,
+		crm:       cloudResourceManager,
 	}, nil
 }
 
-func (p *Provider) ListAllAssetTypesByName(ctx context.Context, assetTypes []string) ([]*ExtendedGcpAsset, error) {
-	wg := sync.WaitGroup{}
-	var resourceAssets []*assetpb.Asset
-	var policyAssets []*assetpb.Asset
-
-	wg.Add(1)
-	go func() {
-		resourceAssets = p.getAllAssets(ctx, &assetpb.ListAssetsRequest{
-			Parent:      p.config.Parent,
-			AssetTypes:  assetTypes,
-			ContentType: assetpb.ContentType_RESOURCE,
-		})
-		wg.Done()
-	}()
-	wg.Add(1)
-	go func() {
-		policyAssets = p.getAllAssets(ctx, &assetpb.ListAssetsRequest{
-			Parent:      p.config.Parent,
-			AssetTypes:  assetTypes,
-			ContentType: assetpb.ContentType_IAM_POLICY,
-		})
-		wg.Done()
-	}()
+func (p *Provider) ListAssetTypes(ctx context.Context, assetTypes []string, out chan<- *ExtendedGcpAsset) {
+	defer close(out)
 
-	wg.Wait()
+	resourceCh := make(chan *assetpb.Asset) // *assetpb.Asset with Resource
+	policyCh := make(chan *assetpb.Asset)   // *assetpb.Asset with IamPolicy
+	mergeCh := make(chan *assetpb.Asset)    // *assetpb.Asset with Resource and IamPolicy
+	enrichCh := make(chan *ExtendedGcpAsset)
 
-	var assets []*assetpb.Asset
-	assets = append(append(assets, resourceAssets...), policyAssets...)
-	mergedAssets := mergeAssetContentType(assets)
-	extendedAssets := p.extendWithCloudMetadata(ctx, mergedAssets)
-	// Enrich network assets with dns policy
-	p.enrichNetworkAssets(ctx, extendedAssets)
+	go p.getAllAssets(ctx, p.config.Parent, assetpb.ContentType_RESOURCE, assetTypes, resourceCh)
+	go p.getAllAssets(ctx, p.config.Parent, assetpb.ContentType_IAM_POLICY, assetTypes, policyCh)
+	go p.mergeAssets(ctx, mergeCh, resourceCh, policyCh)
+	go p.enrichAssets(ctx, mergeCh, enrichCh)
 
-	return extendedAssets, nil
+	for asset := range enrichCh {
+		out <- asset
+	}
```

Comment:
this method is used by the main fetcher (`assets_fetcher`) and the way it works is that we start fetching resources and then merge them by name. the merging process sends assets to the enrichment channel if: 1) an asset has a resource and a policy, or 2) an asset has either a resource or a policy, and the other channel is closed, so we don't need to wait for merging with the other content type

after an asset is enriched it's sent to the out channel which is received in the fetcher and passed to the `resourceCh` to start the relevant flavor (cspm/assets inventory) pipeline